subprojects/platform-native/src/main/java/org/gradle/nativeplatform/TargetMachineFactory.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.nativeplatform;
+
+import org.gradle.api.Incubating;
+
+/**
+ * A factory for creating {@link TargetMachine} objects.
+ *
+ * @since 5.1
+ */
+@Incubating
+public interface TargetMachineFactory {
+    /**
+     * Returns a {@link TargetMachine} representing the operating system and architecture of the current host.
+     */
+    TargetMachine host();
+
+    /**
+     * Returns a {@link TargetMachine} for the windows operating system family.
+     */
+    TargetMachine windows();
+
+    /**
+     * Returns a {@link TargetMachine} for the linux operating system family.
+     */
+    TargetMachine linux();
+
+    /**
+     * Returns a {@link TargetMachine} for the macos operating system family.
+     */
+    TargetMachine macos();
+
+    /**
+     * Returns a {@link TargetMachine} representing the operating system and default architecture.
+     */
+    TargetMachine of(String operatingSystemFamily);
+
+    /**
+     * Returns a {@link TargetMachine} representing the operating system and architecture specified.
+     */
+    TargetMachine of(String operatingSystemFamily, String architecture);
```

Comment:
Hm, there are a couple of crappy parts to this.

- You can't really know which order is valid (since both args are strings) without some documentation.  
- Things like `machines.of('x86')` look valid too.
- the `x86()` and `linux()` etc methods kind of pollute the interfaces.

I guess there are a couple of alternatives...

Type-safe factory methods: 
`TargetMachine TargetMachineFactory.target(OperatingSystemFamilyEnum, ArchitectureEnum)`

It would look something like:
```
targetMachines = [ target(OperatingSystemFamilyEnum.WINDOWS, ArchitectureEnum.X86) ]
```
I think this is the wrong direction.

Fluent builder: 
`TargetMachineBuilder TargetMachineFactory.builder()`
`TargetMachineBuilder.os(String)` 
`TargetMachineBuilder.architecture(String)`, etc.

It would look like 
```
targetMachines = [ machines.builder().os("windows").architecture("x86").build() ]
```
This is promising, but really verbose. We might be able to do some sugaring here? e.g. with a Map
```
targetMachines = [ machines.target(os: "windows", architecture: "x86") ]
```
Maybe there's a built-in Kotlin feature here that would let us do something similar? I've seen something like this in places:
```
targetMachines = [ machines.target(os = "windows", architecture = "x86") ]
```

Similar to the fluent builder above, we could use a config object instead:

`TargetMachine TargetMachineFactory.target(Action<ConfigurableTargetMachine>)`

It would look like 
```
targetMachines = [ 
    machines.target { 
        it.os = "windows"
        it.architecture = "x86"
    } 
]
```
This is sort of promising too, but I worry the nested []/{}s might be offputting.

Reply:
So, I went with a fluent api but not quite a full-scale builder pattern.  This feels pretty good to me.  Basically, it always defaults to the current architecture, but you can be as explicit as you want or mix/match convenience methods:
```
// current os, current architecture
targetMachines = [machines.host()]

// 'foo' os, current architecture
targetMachines = [machines.os('foo')]

// 'foo' os, 'bar' architecture
targetMachines = [machines.os('foo').architecture('bar')]

// 'foo' os, x86 architecture
targetMachines = [machines.os('foo').x86()]

// current os, 'bar' architecture
targetMachines = [machines.host().architecture('bar')]

// linux os, 'bar' architecture
targetMachines = [machines.linux().architecture('bar')]
```

WDYT?